### PR TITLE
feat: capped minter v2 time controls

### PR DESCRIPTION
**Pull request by marcomariscal**

### CODE REVIEW

File: l2-contracts/test/ZkCappedMinterV2.t.sol [hunk 1]
```diff
@@ -12,13 +12,29 @@ contract ZkCappedMinterV2Test is ZkTokenTest {
...
 
+  function _boundToValidTimeControls(uint256 _startTime, uint256 _expirationTime)
+    internal
+    view
+    returns (uint256, uint256)
+  {
+    // Using uint32 for time controls to prevent overflows in the ZkToken contract regarding block numbers needing to be
+    // casted to uint32.
+    _startTime = bound(_startTime, vm.getBlockTimestamp(), type(uint32).max - 1);
+    _expirationTime = bound(_expirationTime, _startTime + 1, type(uint32).max);
```
Comment (alexkeating) [inline l2-contracts/test/ZkCappedMinterV2.t.sol:34]: Could uint48 work?

File: l2-contracts/test/ZkCappedMinterV2.t.sol [hunk 2]
```diff
@@ -37,51 +53,92 @@ contract ZkCappedMinterV2Test is ZkTokenTest {
...
+  }
+
+  function testFuzz_RevertIf_StartTimeAfterExpirationTime(
+    address _admin,
+    uint256 _cap,
+    uint256 _startTime,
+    uint256 _invalidExpirationTime
+  ) public {
+    _cap = bound(_cap, 0, MAX_MINT_SUPPLY);
+    _startTime = bound(_startTime, vm.getBlockTimestamp() + 1, type(uint256).max - 1);
```
Comment (alexkeating) [inline l2-contracts/test/ZkCappedMinterV2.t.sol:80]: I would warp to a random time and then make startTime block.timestamp + 1

Comment (alexkeating) [reply l2-contracts/test/ZkCappedMinterV2.t.sol:80]: And apply a similar idea to the next test

File: l2-contracts/test/ZkCappedMinterV2.t.sol [hunk 3]
```diff
@@ -143,20 +282,76 @@ contract Mint is ZkCappedMinterV2Test {
...
     _cap = bound(_cap, 0, MAX_MINT_SUPPLY);
     vm.assume(_cap > 0);
     _amount = bound(_amount, 1, _cap);
     vm.assume(_admin != address(0));
     vm.assume(_receiver != address(0) && _receiver != initMintReceiver);
+    (_startTime, _expirationTime) = _boundToValidTimeControls(_startTime, _expirationTime);
+    vm.warp(_startTime);
 
-    ZkCappedMinterV2 cappedMinter = createCappedMinter(_admin, _cap);
+    ZkCappedMinterV2 cappedMinter = _createCappedMinter(_admin, _cap, _startTime, _expirationTime);
```
Comment (alexkeating) [inline l2-contracts/test/ZkCappedMinterV2.t.sol:297]: This is an example where maybe moving this to a helper could make things a little cleaner

File: l2-contracts/test/ZkCappedMinterV2.t.sol (hunk 1 above)
Comment (marcomariscal) [reply l2-contracts/test/ZkCappedMinterV2.t.sol:34]: tied, but doesn't work cuz of the block num casting

File: l2-contracts/test/ZkCappedMinterV2.t.sol (hunk 2 above)
Comment (marcomariscal) [reply l2-contracts/test/ZkCappedMinterV2.t.sol]: > I would warp to a random time and then make startTime block.timestamp + 1

i reworked this test to be more clear: https://github.com/ScopeLift/zk-governance/pull/13/commits/d1a730609a37bca210547ee5c81c76ac15b9c17d

lemme know how that works

Comment (marcomariscal) [reply l2-contracts/test/ZkCappedMinterV2.t.sol]: > > I would warp to a random time and then make startTime block.timestamp + 1
> 
> i reworked this test to be more clear: [d1a7306](https://github.com/ScopeLift/zk-governance/commit/d1a730609a37bca210547ee5c81c76ac15b9c17d)
> 
> lemme know how that works

and here actually:

https://github.com/ScopeLift/zk-governance/pull/13/commits/70b49f0eb5af7b3bd1654edef1733f1c777c1583

Comment (marcomariscal) [reply l2-contracts/test/ZkCappedMinterV2.t.sol]: and here i made updates too:

https://github.com/ScopeLift/zk-governance/pull/13/commits/4b8499f34cea5191161b88879c742b237fade253

File: l2-contracts/test/ZkCappedMinterV2.t.sol [hunk 4]
```diff
@@ -10,19 +10,40 @@ import {console2} from "forge-std/Test.sol";
...
+  uint256 DEFAULT_EXPIRATION_TIME;
+
   address cappedMinterAdmin = makeAddr("cappedMinterAdmin");
 
   function setUp() public virtual override {
     super.setUp();
 
-    cappedMinter = _createCappedMinter(cappedMinterAdmin, DEFAULT_CAP);
+    DEFAULT_START_TIME = vm.getBlockTimestamp();
+    DEFAULT_EXPIRATION_TIME = DEFAULT_START_TIME + 100_000_000;
```
Comment (alexkeating) [inline l2-contracts/test/ZkCappedMinterV2.t.sol:22]: Easier to read if you use `1 days` or something

Comment (marcomariscal) [reply l2-contracts/test/ZkCappedMinterV2.t.sol]: https://github.com/ScopeLift/zk-governance/pull/13/commits/083ef5f1a67238f4807ae6034b7ead9f17cbe122

**Review by alexkeating — APPROVED** [review]

**Pull request merged**